File: epi_judge_cpp/is_tree_symmetric.cc
#include "binary_tree_node.h"
#include "test_framework/generic_test.h"

bool IsSymmetricRec(const unique_ptr<BinaryTreeNode<int>> &node1, const unique_ptr<BinaryTreeNode<int>> &node2)
{

  if (node1 == nullptr && node2 == nullptr)
  {
    return true;
  }
  if (node1 == nullptr)
  {
    return false;
  }
  if (node2 == nullptr)
  {
    return false;
  }

  if (node1->data != node2->data)
  {
    return false;
  }

  return IsSymmetricRec(node1->left, node2->right) && IsSymmetricRec(node1->right, node2->left);
}

bool IsSymmetric(const unique_ptr<BinaryTreeNode<int>> &tree)
{
  // you forgot to check for nullptr
  if (tree == nullptr)
  {
    return true;
  }
  return IsSymmetricRec(tree->right, tree->left);
}

int main(int argc, char *argv[])
{
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"tree"};
  return GenericTestMain(args, "is_tree_symmetric.cc", "is_tree_symmetric.tsv",
                         &IsSymmetric, DefaultComparator{}, param_names);
}

File: epi_judge_python/is_tree_symmetric.py
from binary_tree_node import BinaryTreeNode
from test_framework import generic_test

# what is the issue here
def is_symmetric(tree: BinaryTreeNode) -> bool:

    def is_sym(left,right):
        if left is None and right is None:
            return True
        if left is None:
            return False
        if right is None:
            return False
        

        if left.data != right.data:
            return False
        
        else:
            # needs to be mirrored
            return is_sym(left.left,right.right) and is_sym(left.right,right.left)

        
        
    if tree is None:
        return True
    
    return is_sym(tree.left,tree.right)


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('is_tree_symmetric.py',
                                       'is_tree_symmetric.tsv', is_symmetric))
